### PR TITLE
Use getBoundingRects to add support inset MediaQuery/SafeArea when in freeform mode controls are shown. 

### PR DESCRIPTION
**Pull request by yaakovschectman**
Original Title: Add FlutterViewDelegate and BoundingRect methods

Check the bounding rect for caption bar when sending viewport metrics to Flutter to account for freeform mode. Use the more recent `getBoundingRects` over `getInsets`.

Tests will need to be kept separate until Robolectric publishes a version that supports API level 35.
Roboletric tests for this pr will be part of google testing until then. 
Pr for tests here https://critique.corp.google.com/cl/657302386. 

https://github.com/flutter/flutter/issues/146658

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java
Comment (gmackall) [inline shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:31]: `getWindow()` and `getRootWindowInsets()` can both return null. Looking at existing code it looks like it's a mixed bag on whether these cases have historically been handled. 

I'm not sure we could currently run in to those cases, but it would be good to handle in a general delegate class to make sure we don't run into any `NPE`s in the future.

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java
Comment (gmackall) [inline shell/platform/android/io/flutter/embedding/android/FlutterView.java:1454]: nit: I think this comment can be removed (it setting is intuitive, it being used for testing is implied by `@VisibleForTesting`). If you wanted to keep it, I'd be more explicit ("... only used for testing").

File: shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java
Comment (gmackall) [inline shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:23]: nit: Comment is out of date, returns null when any of
1. no associated activity
2. the window is null
3. `getRootWindowInsets()` returns null

**Review by gmackall — APPROVED** [review]
LGTM outside of nit

**Comment by chinmaygarde** [comment]
Is this blocked on anything? Looks good to go.

**Comment by reidbaker** [comment]
> Is this blocked on anything? Looks good to go.

It is waiting on my review. I was reviewing the internal cl which has the test but the files diverged.

File: shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:41]: This annotation is incorrect. This is used publicly in 1472 in FlutterView. If you want to reduce the scope of this method maybe package private would but not @VisibleForTesting would be correct. 

@gmackall, I am surprised this passes our java lint, something for us to investigate later.

File: shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:43]: It looks like we should be moving to window insets for all android code after api 30. Not in this pr but in general. Based on your understanding of this android code is that correct?

File: shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:40]: Nothing in this method is limited to api 35. it appears WindowInset.Type.captionBar is available in api 30. 

Am I missing something that is limited to api 35?

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/android/FlutterView.java:1471]: If the other class changes getCaptionBarInsets then this api check should also change.

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/android/FlutterView.java:1455]: Add non null annotation here.

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/android/FlutterView.java:1459]: Consider adding a link to https://github.com/flutter/engine/blob/main/lib/ui/hooks.dart#L139-L155 in the ViewportMetrics class (not here) so that future callers can more easily find how it is used.

**Review by reidbaker — CHANGES_REQUESTED** [review]

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java
Comment (yaakovschectman) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java:1473]: I had assumed the caption bar would have only one rect. If this is false, what would you suppose we do if there are more, considering that the viewport metrics describe only a bounding rect for the view? Get the maximum to use for the padding value?

File: shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java
Comment (yaakovschectman) [reply shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:43]: I don't understand your question. What do you mean by "moving to window insets"? Moving what, and from where?

File: shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java
Comment (yaakovschectman) [reply shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:40]: `getBoundingRects` is called here, and is new.

File: shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java
Comment (yaakovschectman) [reply shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:22]: Can you explain why this detail is relevant documentation to this method? I'm not sure I see the throughline.

File: shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java
Comment (reidbaker) [reply shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:41]: Facepalm, sorry that is correct.

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java
Comment (reidbaker) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java]: We can sometimes work from assumptions. I am asking what evidence justifies ignoring that the api returns a list. I dont know if bounding recs are a single value or list. That is your responsibility as the author of a change.

File: shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java
Comment (reidbaker) [reply shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:27]: The dart docs is a hint for callers and for maintainers to help understand the behavior of the code documented. The existing dart doc, as written, helps a little bit by describing the null behavior but it could be more helpful if it describe when context might change. That is because the caller can know for their use case of FlutterViewDelegate if it matters which context to pass.

For example if the android documentation indicates that window insets will always be the same for an app regardless of which of the apps context is used it indicated to the caller that it might be safe in situations when launching another activity to use the application context. However if insets might change across activities maybe because of their window flags then it is important to use one FlutterViewDelate per activity.

File: shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java
Comment (reidbaker) [reply shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:47]: Let me try to rephrase: 
Do you, as the person who most recently read and worked with the window insets and its documentation, think that on api 30+ we need to rewrite any of the code that sets ViewportMetrics to ensure that ViewportMetrics is getting the correct values from the android embedder?

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/android/FlutterView.java:1459]: Looking at the design of this class I am seeing a pattern of the flutterView having methods that trigger on different android lifecycle events, Then ViewportMetrics being updated then at the end of that triggering method calling `sendViewportMetricsToFlutter`. 

Looking at this code we do the calculation every send. When I was searching internally I did not find documentation of a lifecycle method to trigger on so maybe this is moot. If this code moved to onApplyWindowInsets would it still work?

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/android/FlutterView.java:1]: Digging into [go/customizable-window-headers](http://goto.google.com/customizable-window-headers) I found [APPEARANCE_TRANSPARENT_CAPTION_BAR_BACKGROUND](https://developer.android.com/reference/kotlin/android/view/WindowInsetsController.html#APPEARANCE_TRANSPARENT_CAPTION_BAR_BACKGROUND:kotlin.Int) Can you test an app with that flag set vs the default and come up with a recommendation on if we should apply transparent caption bar by default or not?

Comment (reidbaker) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java:1]: This comment should not be considered blocking.

File: shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java
Comment (yaakovschectman) [reply shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:47]: The most comprehensive response I can give is that we currently already use the `WindowInsets` API, but the older inset getter method rather than the newer bounding rects method. Switching to the newer method in the future would allow us to cover more robust scenarios, such as where the safe area of a view is not a simple rectangle, but this would also require updating the implementation of `SafeArea`, and it looks to me like that update would constitute a breaking change.

Comment (reidbaker) [reply shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:47]: Ok so either a breaking change to safe area or a different new widget that let items grow in irregular shapes. Thank you, for now I think this means we dont need a rewrite and I appreciate the answer.

File: shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:30]: Great documentation

File: shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:64]: I think based on this new code you want to remove this size check.

File: shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/android/FlutterViewDelegate.java:59]: Consider naming padding something that more closely aligns with the variable here. Padding is used in a few places in flutterView.

**Review by reidbaker — APPROVED** [review]

**Comment by reidbaker** [comment]
@mengyanshou yaakov no longer contributes to flutter it is unlikely you will get a response. Please file an issue so that the triage process can start.

**Pull request merged**
